File: backend/app/sensors/thermal.py
from __future__ import annotations

import numpy as np

from app.core.voxel_engine import VoxelGrid
from app.sensors.base import BaseSensorSimulator, local_contrast, surface_skin_weight


class ThermalSimulator(BaseSensorSimulator):
    key = "thermal"
    name_tr = "Termal / Kızılötesi"
    name_en = "Thermal Infrared"
    description_tr = "Sığ gömülü yapıların yüzey ısı akışını değiştirmesiyle oluşan yüzey sıcaklık anomalilerini görüntüler."
    icon = "thermal"
    max_effective_depth_m = 1.0
    base_reliability = 0.55

    def _raw_signal(self, grid: VoxelGrid) -> np.ndarray:
        conductivity = grid.property_grid("thermal_conductivity")
        contrast = local_contrast(conductivity)
        weight = surface_skin_weight(grid, skin_depth_m=0.5)
        return contrast * weight[None, None, :]

    def profile_2d(self, grid: VoxelGrid, volume: np.ndarray):
        weight = np.arange(grid.nz, 0, -1, dtype=float)
        surface_map = np.tensordot(volume, weight, axes=([2], [0])) / weight.sum()
        return surface_map, ("x", "y")

File: frontend/src/sim/sensors/thermal.ts
import type { VoxelGrid } from "../voxelEngine";
import { localContrast, surfaceSkinWeight, type SensorDef } from "./base";

export const ThermalSensor: SensorDef = {
  key: "thermal",
  name_tr: "Termal / Kızılötesi",
  name_en: "Thermal Infrared",
  description_tr: "Sığ gömülü yapıların yüzey ısı akışını değiştirmesiyle oluşan yüzey sıcaklık anomalilerini görüntüler.",
  icon: "thermal",
  max_effective_depth_m: 1.0,
  base_reliability: 0.55,

  rawSignal(grid: VoxelGrid): Float64Array {
    const conductivity = grid.propertyGrid("thermal_conductivity");
    const contrast = localContrast(grid, conductivity);
    const weight = surfaceSkinWeight(grid, 0.5);
    const out = new Float64Array(contrast.length);
    for (let x = 0; x < grid.nx; x++) {
      for (let y = 0; y < grid.ny; y++) {
        for (let z = 0; z < grid.nz; z++) {
          const i = grid.idx(x, y, z);
          out[i] = contrast[i] * weight[z];
        }
      }
    }
    return out;
  },

  profile2D(grid: VoxelGrid, volume: Float64Array) {
    const weight: number[] = [];
    for (let z = 0; z < grid.nz; z++) weight.push(grid.nz - z);
    const weightSum = weight.reduce((a, b) => a + b, 0);
    const data: number[][] = [];
    for (let x = 0; x < grid.nx; x++) {
      const row: number[] = [];
      for (let y = 0; y < grid.ny; y++) {
        let sum = 0;
        for (let z = 0; z < grid.nz; z++) sum += volume[grid.idx(x, y, z)] * weight[z];
        row.push(sum / weightSum);
      }
      data.push(row);
    }
    return { data, axes: ["x", "y"] as [string, string] };
  },
};
